### PR TITLE
[OpenCL] Add bounds checking to the Enqueue memory operations.

### PR DESCRIPTION
**Pull request by aarongreig**
This allows us to return UR_ERROR_INVALID_SIZE when we should. Extra
checks are only performed on a non-success error code.

### CODE REVIEW

**Comment by aarongreig** [comment]
I had to base this on the mega fix branch as there was a non-trivial overlap in EnqueueMemBufferFill. Only the top commit is relevant.

**Comment by aarongreig** [comment]
(and by top commit I mean most recent... which on github is the bottom commit)

**Review by kbenzie — APPROVED** [review]
LGTM

File: source/adapters/opencl/device.cpp [hunk 1]
```diff
@@ -345,6 +345,23 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGetInfo(ur_device_handle_t hDevice,
...
+
+    if (!Supported) {
+      return UR_RESULT_ERROR_UNSUPPORTED_ENUMERATION;
+    }
+
+    cl_device_pci_bus_info_khr PciInfo = {};
+    CL_RETURN_ON_FAILURE(clGetDeviceInfo(
+        cl_adapter::cast<cl_device_id>(hDevice), CL_DEVICE_PCI_BUS_INFO_KHR,
+        sizeof(PciInfo), &PciInfo, nullptr));
+    return ReturnValue(PciInfo.pci_device);
```
Comment (omarahmed1111) [inline source/adapters/opencl/device.cpp:362]: I think we could check here for the version of opencl first as I think this query should be only available from version 3.0, I guess we could guard that behind a preprocessor if, not sure if this should work in a preprocessor or not but there is `UR_EXT_DEVICE_INFO_OPENCL_C_VERSION`define in `plugins/unified_runtime/ur/ur.hpp`

**Review by omarahmed1111 — APPROVED** [review]
LGTM, just a small nit

File: source/adapters/opencl/device.cpp (hunk 1 above)
Comment (aarongreig) [reply source/adapters/opencl/device.cpp:362]: I think regardless of underlying driver version we don't support building with a CL loader/headers that would be old enough not to have this, if it isn't supported by the device that should get caught by the check for cl_khr_pci_bus_info above

Comment (omarahmed1111) [reply source/adapters/opencl/device.cpp:362]: ah okay, that make sense. I thought we could as I saw cmake searching for cl version 1.0 and 2.0 before moving to fetching headers from upstream.

**Pull request closed without merging**